### PR TITLE
[WIP] Replace flying saucer with open html to pdf

### PR DESCRIPTION
**Pull request by tiz170**
This commit replaces flying saucer with open html to pdf to generate pdf from html. 

With this code, things that work:
- generates pdfs with images successfully
- supports image orientation

things that do not work:
- there is no table of content in the generated pdf
- logo is missing from the generated pdf

I tried to add `pdfRendererBuilder.usePdfUaAccessbility(true)` but it throws `java.lang.IndexOutOfBoundsException` excxeption. Still debugging.

### CODE REVIEW

File: src/main/scala/org/orbeon/oxf/processor/pdf/XHTMLToPDFProcessor.scala [hunk 1]
```diff
@@ -78,57 +89,87 @@ class XHTMLToPDFProcessor() extends HttpBinarySerializer {
...
+          // tiz170 TODO: add OHTP correspondence of creating PDF
+          pdfBoxRenderer.createPDF()
+        }
       }
+
     } finally {
       // Free resources associated with the rendering context
-      renderer.getSharedContext.reset()
+      //      renderer.getSharedContext.reset()
     }
```
Comment (ebruchez) [inline src/main/scala/org/orbeon/oxf/processor/pdf/XHTMLToPDFProcessor.scala:118]: Here you will have to call probably the `.close()` method on `pdfBoxRenderer` to free the resources in the `finally` block.

File: src/main/resources/config/properties-form-runner.xml [hunk 2]
```diff
@@ -240,6 +240,8 @@
     <property as="xs:string"  name="oxf.fr.pdf.font.path.ID"                                    value=""/>
     <property as="xs:string"  name="oxf.fr.pdf.font.family.ID"                                  value=""/>
     -->
+    <property as="xs:string"  name="oxf.fr.pdf.font.resource.roboto"                            value="/forms/resources/Roboto-Medium.ttf"/>
+    <property as="xs:string"  name="oxf.fr.pdf.font.family.roboto"                              value="Roboto"/>
     <property as="xs:string" name="oxf.fr.pdf.template.font.paths"                              value=""/>
```
Comment (ebruchez) [inline src/main/resources/config/properties-form-runner.xml:246]: We will have to check if/why specifying an explicit font to embed is required. It's unclear to me why it's the case. Also I expect the behavior to be that this means some CSS refers to the `roboto` font, and I couldn't find an example. This just to say that we'll have to dig into this a little more.

Comment (tiz170) [reply src/main/resources/config/properties-form-runner.xml:246]: Yeah I was looking at this accessibility sample https://github.com/danfickle/openhtmltopdf/wiki/PDF-Accessibility-(PDF-UA,-WCAG,-Section-508)-Support where it has `useFont` and it says `Remember to add one or more fonts.` which makes me think this is necessary. However it turns out this is not the cause of failure. Will remove.

File: src/main/scala/org/orbeon/oxf/processor/pdf/XHTMLToPDFProcessor.scala (hunk 1 above)
Comment (tiz170) [reply src/main/scala/org/orbeon/oxf/processor/pdf/XHTMLToPDFProcessor.scala:118]: Thanks. Added.

File: src/main/scala/org/orbeon/oxf/processor/pdf/XHTMLToPDFProcessor.scala [hunk 3]
```diff
@@ -78,57 +89,87 @@ class XHTMLToPDFProcessor() extends HttpBinarySerializer {
...
+      //        )
+      //      )
+
+      ////       tiz170: this is where data is passed into flying-saucer
+      //      renderer.setDocument(
+      //        readInputAsDOM(pipelineContext, input),
+      //        requestOpt map (_.getRequestURL) orNull // no base URL if can't get request URL from context
+      //      )
+      //      renderer.layout()
+
```
[the remark this answers is not in the document]
Comment (tiz170) [reply src/main/scala/org/orbeon/oxf/processor/pdf/XHTMLToPDFProcessor.scala]: I tested an image that has orientation and it's not rendered correctly so looks like we still need to manually handle this. Will look into this.

[the remark this answers is not in the document]
Comment (tiz170) [reply src/main/scala/org/orbeon/oxf/processor/pdf/XHTMLToPDFProcessor.scala]: I override `getImageResource` and add code to support image orientation. Now the image that has orientation is rendered correctly.

**Comment by tiz170** [comment]
Hey @ebruchez I can see `PdfBoxBookmarkManager` writes outline by writing bookmarks ([writeOutline](https://github.com/danfickle/openhtmltopdf/blob/ccd29f03ede2aecadac9c39fda95a5fedfb23645/openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxBookmarkManager.java#L46)), and bookmarks are coming from `bookmark` element in `head` or `body` nodes ([loadBookmarks](https://github.com/danfickle/openhtmltopdf/blob/ccd29f03ede2aecadac9c39fda95a5fedfb23645/openhtmltopdf-pdfbox/src/main/java/com/openhtmltopdf/pdfboxout/PdfBoxBookmarkManager.java#L149)). 
However I checked the input Document that's parsed from `readInputAsDOM(pipelineContext, input)` and passed to `pdfRendererBuilder` at the very beginning and there is no `bookmark` element in the raw HTML at all.  See below for a sample raw HTML:
```
<html id="xf-1" lang="en" xml:lang="en">
    <head>
        <META http-equiv="Content-Type" content="text/html; charset=UTF-8">
        <title xmlns:frf="java:org.orbeon.oxf.fr.FormRunner" xmlns:p="http://www.orbeon.com/oxf/pipeline" xmlns:saxon="http://saxon.sf.net/" xmlns:xf="http://www.w3.org/2002/xforms" xmlns:xh="http://www.w3.org/1999/xhtml" xmlns:xs="http://www.w3.org/2001/XMLSchema" id="xf-6">Untitled Form</title>
        <link href="/orbeon/xforms-server/orbeon-467b616441ef81185c851f2170b05d2f8cc4e721.css" media="all" rel="stylesheet" type="text/css">
        <link href="/orbeon/xforms-server/orbeon-0ffa30f9d0aff64824b51cb3db197f04500cae2e.css" media="all" rel="stylesheet" type="text/css">
        <style type="text/css">
                @page {
                    size: letter portrait;
                }
            </style>
    </head>
    <body class="orbeon xforms-disable-alert-as-tooltip yui-skin-sam">
        <div class="fr-header-title xforms-hidden">
            <div>Untitled Form</div>
        </div>
        <div class="fr-footer-title xforms-hidden">
            <div>Untitled Form</div>
        </div>
        <form action="/orbeon/fr/service/123/123/test-pdf" class="xforms-hint-appearance-full xforms-form xforms-label-appearance-full xforms-help-appearance-dialog xforms-disable-hint-as-tooltip" id="xforms-form" method="POST" onsubmit="return false">
            <div class="container fr-view fr-mode-pdf xforms-group" id="fr-view">
                <div class="navbar navbar-inverse xforms-group xforms-readonly" id="xf-295">
                    <div class="navbar-inner">
                        <div class="container">
                            <div>
                                <img alt="" src="/orbeon/apps/fr/style/orbeon-navbar-logo.png">
                            </div>
                            <h1>
                                <span class="xforms-control xforms-output" id="xf-301"><span class="xforms-output-output" id="xf-301&equiv;&equiv;c">Untitled Form</span></span>
                            </h1>
                            <div class="fr-language-choice xforms-group" id="xf-302"></div>
                            <span class="fr-status-icons xforms-group xforms-readonly" id="xf-304"><span class="fr-validity-icon xforms-group" id="xf-310"><span class="xforms-group" id="xf-311"><i class="fa fa-check fa-fw" id="xf-312" title="No errors on form"></i></span></span></span>
                        </div>
                    </div>
                </div>
                <div class="row">
                    <div class="span12"></div>
                </div>
                <div class="row">
                    <div class="span12"></div>
                </div>
                <div class="row">
                    <div class="span12">
                        <span class="fr-body fr-border fr-validation-mode-incremental xforms-group xforms-readonly" id="fr-form-group"><a name="fr-form"></a><span class="fr-view-appearance-full xforms-group" id="fr-view-component"> 
                                <div class="xforms-readonly xbl-component xbl-fr-section" id="section-1-section">
                                    <div class="xforms-group" id="section-1-section&equiv;xf-608">
                                        <h2 class="fr-section-title">
                                            <span class="fr-section-label fr-section-open xforms-control xforms-trigger xforms-trigger-appearance-minimal" id="section-1-section&equiv;xf-614"><button class="xforms-help"></button><button class="btn-link" id="section-1-section&equiv;xf-614&equiv;&equiv;c" title="">section 1</button></span>
                                        </h2>
                                        <div class="xforms-update-full xforms-switch" id="section-1-section&equiv;switch">
                                            <span class="xforms-case-selected">
                                                <div class="fr-section-content xxforms-animate xforms-group" id="section-1-section&equiv;xf-618">
                                                    <div class="xforms-readonly xbl-component xbl-fr-grid-single xbl-fr-grid" id="section-1-section&equiv;grid-1-grid">
                                                        <table class="fr-grid fr-grid-2 fr-grid-grid-1 fr-norepeat" role="presentation">
                                                            <tbody class="fr-grid-body">
                                                                <tr class="fr-grid-tr">
                                                                    <td class="fr-grid-td xforms-group" id="section-1-section&equiv;grid-1-grid&equiv;xf-631"><span class="fr-grid-1-1 xforms-control xforms-input xforms-static xforms-readonly" id="section-1-section&equiv;grid-1-grid&equiv;control-1-control"><span class="xforms-label">test</span><span class="xforms-field"><span>test content</span></span></span></td><td class="fr-grid-td xforms-group" id="section-1-section&equiv;grid-1-grid&equiv;xf-636"></td>
                                                                </tr>
                                                            </tbody>
                                                        </table>
                                                        <div xmlns:xxbl="http://orbeon.org/oxf/xml/xbl" class="fr-grid-non-empty" id="section-1-section&equiv;grid-1-grid&equiv;xf-638" xxbl:scope="outer"></div>
                                                    </div>
                                                </div>
                                            </span>
                                        </div>
                                    </div>
                                </div>
                                <div class="xforms-readonly xbl-component xbl-fr-section" id="section-2-section">
                                    <div class="xforms-group" id="section-2-section&equiv;xf-643">
                                        <h2 class="fr-section-title">
                                            <span class="fr-section-label fr-section-open xforms-control xforms-trigger xforms-trigger-appearance-minimal" id="section-2-section&equiv;xf-649"><button class="xforms-help"></button><button class="btn-link" id="section-2-section&equiv;xf-649&equiv;&equiv;c" title="">2</button></span>
                                        </h2>
                                        <div class="xforms-update-full xforms-switch" id="section-2-section&equiv;switch">
                                            <span class="xforms-case-selected">
                                                <div class="fr-section-content xxforms-animate xforms-group" id="section-2-section&equiv;xf-653">
                                                    <div class="xforms-readonly xbl-component xbl-fr-grid-single xbl-fr-grid" id="section-2-section&equiv;grid-2-grid">
                                                        <table class="fr-grid fr-grid-2 fr-grid-grid-2 fr-norepeat" role="presentation">
                                                            <tbody class="fr-grid-body">
                                                                <tr class="fr-grid-tr">
                                                                    <td class="fr-grid-td xforms-group" id="section-2-section&equiv;grid-2-grid&equiv;xf-666"><span class="fr-grid-1-1 xforms-control xforms-input xforms-static xforms-readonly" id="section-2-section&equiv;grid-2-grid&equiv;control-2-control"><span class="xforms-label">test2</span><span class="xforms-field"><span>test2 content</span></span></span></td><td class="fr-grid-td xforms-group" id="section-2-section&equiv;grid-2-grid&equiv;xf-671"></td>
                                                                </tr>
                                                            </tbody>
                                                        </table>
                                                        <div class="fr-grid-non-empty" id="section-2-section&equiv;grid-2-grid&equiv;xf-673" xxbl:scope="outer"></div>
                                                    </div>
                                                </div>
                                            </span>
                                        </div>
                                    </div>
                                </div>
                                 </span></span>
                    </div>
                </div>
                <div class="row">
                    <div class="span12">
                        <div class="xbl-component xbl-fr-captcha" id="fr-captcha">
                            <span class="xforms-group" id="fr-captcha&equiv;xf-676"></span>
                        </div>
                    </div>
                </div>
                <div>
                    <row>
                        <div class="xbl-component xbl-fr-error-summary" id="error-summary-control-bottom">
                            <span aria-live="polite"></span>
                        </div>
                    </row>
                </div>
                <div class="row">
                    <div class="span12"></div>
                </div>
                <div class="row">
                    <div class="span12">
                        <span class="fr-buttons xforms-group xforms-readonly" id="xf-449"><span class="fr-buttons-message">
                                <div class="xforms-control xforms-output xforms-mediatype-text-html xforms-mediatype-text xforms-static xforms-readonly" id="xf-455">
                                    <div class="xforms-output-output" id="xf-455&equiv;&equiv;c"></div>
                                </div>
                            </span><span class="xforms-repeat-begin-end" id="repeat-begin-xf-457"></span><span class="xforms-repeat-begin-end" id="repeat-end-xf-457"></span></span>
                    </div>
                </div>
                <div class="row">
                    <div class="span12"></div>
                </div>
                <div class="xbl-component xbl-fr-alert-dialog" id="fr-error-dialog">
                    <div class="xbl-fr-alert-dialog xbl-fr-alert-dialog-info xforms-dialog xforms-dialog-appearance-full xforms-initially-hidden xforms-dialog-modal xforms-dialog-close-false xforms-dialog-draggable-true xforms-dialog-visible-false" id="fr-error-dialog&equiv;dialog">
                        <div class="hd xxforms-dialog-head">Unable to complete action</div>
                        <div class="bd xxforms-dialog-body">
                            <div class="xbl-fr-alert-dialog-message"></div>
                            <div class="xbl-fr-alert-dialog-buttons"></div>
                        </div>
                    </div>
                </div>
                <div class="xbl-component xbl-fr-alert-dialog" id="fr-confirmation-dialog">
                    <div class="xbl-fr-alert-dialog xbl-fr-alert-dialog-question xforms-dialog xforms-dialog-appearance-full xforms-initially-hidden xforms-dialog-modal xforms-dialog-close-false xforms-dialog-draggable-true xforms-dialog-visible-false" id="fr-confirmation-dialog&equiv;dialog">
                        <div class="hd xxforms-dialog-head">Confirmation</div>
                        <div class="bd xxforms-dialog-body">
                            <div class="xbl-fr-alert-dialog-message"></div>
                            <div class="xbl-fr-alert-dialog-buttons"></div>
                        </div>
                    </div>
                </div>
            </div>
        </form>
    </body>
</html>

```
Does this mean we need to update the code to manually add bookMark elements to the HTML Document so that they can be consumed by `PdfBoxBookmarkManager`?

Btw, good news is I found this [github repo](https://github.com/kaleguy/openhtmltopdf-dynamic-html/blob/master/src/main/java/com/saltroadapp/leopdf/LeoPDF.java) which uses OHTP to convert HTML to pdf and has a sample input and output. I can see as long as the input HTML has `bookmark` elements, the output pdf will have the "table of content"/outline we are looking for.

**Comment by ebruchez** [comment]
My understanding is that bookmarks are related to, but different from the outline. That is, you can have bookmarks that don't belong to the outline. But it looked like the outline was using, at least internally, the concept of bookmarks, but I am not sure. (Interesting to see that `<bookmark>` doesn't seem to be a standard HTML element at all.)

For the outline, I'd expect the `<h1>`, etc. hierarchy to be used.

**Comment by avernet** [comment]
Steps 

- [ ] Alex: finish XSLT generating bookmarks
- [ ] Alex: if a font is required for the PDF/UA support, find a TTF for Helvetica Neue that we can ship with Orbeon Forms, maybe from Google Fonts
- [ ] Hollie: remove Flying Saucer from `build.sbt`, remaining code using Flying Saucer
- [ ] Hollie: clean up code (remove commented code we don't need anymore, handle comments added during implementation, etc)
- [ ] Hollie: investigate/solve issue with missing logo (see if some of the messages going to the console can help with this)
- [ ] Hollie: see if we can configure OHTP not to output those message to console (or even better if we can route them to our logger)

**Comment by avernet** [comment]
@tiz170 As discussed this afternoon, I've done the 2 steps mentioned above, so you might want to pull the latest changes on this PR before you continue. With this you should now have the table of contents work just fine "out-of-the-box", i.e. without having to add any fonts.

**Comment by tiz170** [comment]
Thanks Alex. 
- [x]  Hollie: remove Flying Saucer from build.sbt, remaining code using Flying Saucer
- [x]  Hollie: clean up code (remove commented code we don't need anymore, handle comments added during implementation, etc)
- [ ]  Hollie: investigate/solve issue with missing logo (see if some of the messages going to the console can help with this)
- [ ]  Hollie: see if we can configure OHTP not to output those message to console (or even better if we can route them to our logger)

**Comment by tiz170** [comment]
Okay identified the root case of the logo image not shown - something is wrong with this image itself. As shown in the screenshot, it fails this check `Can't handle truecolor with alpha, would need to separate alpha from image data`.  
logo_debug

I tried to replace the local image `orbeon-navbar-logo.png` with some other test image and it can be shown correctly at the place of the logo.

**Comment by avernet** [comment]
Next, all for Alex:

- [ ] Check if I can find another PNG for the logo that doesn't cause a problem to OHTP
- [ ] Merge PR
- [ ] Test the accessibility of the generated PDF

**Comment by ebruchez** [comment]
Closing as we have merged the changes on master.

**Pull request closed without merging**